### PR TITLE
simplify operator boundary condition handling

### LINKED ISSUE #1173: GPU Jacobian computation
Confirm that our existing approach works on GPUs.

When I tried this as part of #1200, I hit some issues with determining the boundary width.

### PR DESCRIPTION
**Pull request by simonbyrne**
I needed to rework the operator boundary condition logic to get the Jacobian stencil computation to work on GPUs.

Fixes #1173 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

File: test/Operators/finitedifference/opt.jl [hunk 1]
```diff
@@ -237,7 +237,7 @@ end
             @test_opt opt_LeftBiasedF2C(faces)
             @test_opt opt_RightBiasedF2C(faces)
 
-            @test_opt opt_AdvectionF2F(face_velocities, faces)
+            # @test_opt opt_AdvectionF2F(face_velocities, faces)
```
Comment (charleskawczynski) [inline test/Operators/finitedifference/opt.jl:240]: Can/should we uncomment?

File: test/Operators/hybrid/opt.jl [hunk 2]
```diff
@@ -268,7 +268,7 @@ end
             @test_opt opt_LeftBiasedF2C(faces)
             @test_opt opt_RightBiasedF2C(faces)
 
-            @test_opt opt_AdvectionF2F(face_velocities, faces)
+            # @test_opt opt_AdvectionF2F(face_velocities, faces)
```
Comment (charleskawczynski) [inline test/Operators/hybrid/opt.jl:271]: Can/should we uncomment?

File: test/Operators/finitedifference/implicit_stencils_utils.jl [hunk 3]
```diff
@@ -71,7 +71,9 @@ function get_space(::Type{FT}) where {FT}
         boundary_tags = (:bottom, :top),
     )
     vmesh = Meshes.IntervalMesh(vdomain, nelems = velem)
-    vspace = Spaces.CenterFiniteDifferenceSpace(vmesh)
+    vtopology =
+        Topologies.IntervalTopology(ClimaComms.SingletonCommsContext(), vmesh)
```
Comment (charleskawczynski) [inline test/Operators/finitedifference/implicit_stencils_utils.jl:75]:
```suggestion
    vtopology =
        Topologies.IntervalTopology(ClimaComms.SingletonCommsContext(device), vmesh)
```
?

File: src/Operators/pointwisestencil.jl [hunk 4]
```diff
@@ -337,23 +360,75 @@ get_boundary(
...
+    bc::RightStencilBoundary,
+    stencil,
+    arg,
+)
+    rbw = bandwidths(eltype(stencil))[2]
+    if rbw isa Integer
+        return right_idx(space) - rbw
+    elseif rbw isa PlusHalf
+        return right_idx(space) - rbw + half
+    end
```
Comment (charleskawczynski) [inline src/Operators/pointwisestencil.jl:431]: Should we define

```julia
right_interior_idx(rbw::Integer, space) = right_idx(space) - rbw
right_interior_idx(rbw::PlusHalf, space) = right_idx(space) - rbw + half
```
and use that here and above? Actually, it looks like the two `right_interior_idx` can be unified with these changes?

**Review by dennisYatunin — APPROVED** [review]

File: test/Operators/finitedifference/opt.jl (hunk 1 above)
Comment (simonbyrne) [reply test/Operators/finitedifference/opt.jl:240]: So it turns out that `AdvectionF2F` as it is currently defined doesn't work (i.e. it doesn't have well-defined boundary conditions)

File: test/Operators/finitedifference/implicit_stencils_utils.jl (hunk 3 above)
Comment (simonbyrne) [reply test/Operators/finitedifference/implicit_stencils_utils.jl:75]: `device` isn't defined anywhere?

File: src/Operators/pointwisestencil.jl (hunk 4 above)
Comment (simonbyrne) [reply src/Operators/pointwisestencil.jl:431]: Good idea.

File: test/Operators/finitedifference/opt.jl (hunk 1 above)
Comment (simonbyrne) [reply test/Operators/finitedifference/opt.jl:240]: (and consequently, it now breaks a bunch of tests)

File: test/Operators/finitedifference/implicit_stencils_utils.jl (hunk 3 above)
Comment (charleskawczynski) [reply test/Operators/finitedifference/implicit_stencils_utils.jl:75]: Can't we compute it? I was prompted to suggest this based on the block below:

```julia
        if ClimaComms.device(a_FS) isa ClimaComms.CUDADevice
            @test_throws "InvalidIRError" OP.Operator2Stencil(op).(a)
        else
            @test_throws ArgumentError OP.Operator2Stencil(op).(a)
        end
```

File: test/Operators/finitedifference/opt.jl (hunk 1 above)
Comment (charleskawczynski) [reply test/Operators/finitedifference/opt.jl:240]: Ok, I don't see it used in ClimaAtmos, fine with commenting

File: test/Operators/finitedifference/implicit_stencils_utils.jl (hunk 3 above)
Comment (simonbyrne) [reply test/Operators/finitedifference/implicit_stencils_utils.jl:75]: It's determined automatically (if a GPU is available or not)

Comment (charleskawczynski) [reply test/Operators/finitedifference/implicit_stencils_utils.jl:75]: Ah, I thought we specified the device above at `htopology = Topologies.Topology2D(ClimaComms.SingletonCommsContext(), hmesh)`. This is fine.
